src/bucket/BucketManagerImpl.cpp

```diff
@@ -435,6 +615,77 @@ BucketManagerImpl::forgetUnreferencedBuckets()
                 auto gzfilename = filename + ".gz";
                 std::remove(gzfilename.c_str());
             }
+
+            // Work through the map entries used for resynthesis of a
+            // std::shared_future containing the bucket, removing all the
+            // entries that could have produced it.
+            auto const& outputBeingDropped = j->first;
+            auto mergesProducingOutput =
+                mSharedBucketMergesByOutput.equal_range(outputBeingDropped);
+            for (auto mergeProducingOutput = mergesProducingOutput.first;
+                 mergeProducingOutput != mergesProducingOutput.second;
+                 mergeProducingOutput =
+                     mSharedBucketMergesByOutput.erase(mergeProducingOutput))
+            {
+                auto const& output = mergeProducingOutput->first;
+                auto const& mergeKeyProducingOutput =
+                    mergeProducingOutput->second;
+                assert(output == outputBeingDropped);
+
+                // There should be no futures alive with this output: we
+                // switched to storing only weak input/output mappings when any
+                // merge producing the bucket completed (in adoptFileAsBucket),
+                // and we believe there's only one reference to the bucket
+                // anyways -- our own in mSharedBuckets. But there might be a
+                // race we missed, so double check here.
+                assert(mLiveFutures.find(mergeKeyProducingOutput) ==
```

Comment:
I normally prefer to not have asserts in critical code like this, maybe we should instead just leak memory and buckets? This would make it highly unlikely that all nodes crash at the same time if there is indeed some bug.

On startup, we would clean things up anyways.

Reply:
Fair. Kinda scared me too.